youtrack: 2022.3.65371 -> 2023.3.23390

leona-ya opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

https://www.jetbrains.com/youtrack/whatsnew/

Since JetBrains has deprecated JAR packaging (we used that),
I changed the packaging to use the included ZIP (JetBrains Launcher), renamed the youtrack package to youtrack_2022_3 and add a new package youtrack. 
The update to 2023.x once requires an update with manual interaction. All updates after that will happen automatically.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

leona-ya commented:
I'm not really happy with the solution, but I don't have a better one. Also not 100% sure if the configuration options really work that way as the docs confuse me.

dasJ commented on pkgs/servers/jetbrains/youtrack/default.nix:
```diff
@@ -0,0 +1,85 @@
+
+  generic = {
+    version, hash,
+  }: stdenv.mkDerivation rec {
```
Can also be `stdenvNoCC` here

dasJ commented on pkgs/servers/jetbrains/youtrack/default.nix:
```diff
@@ -0,0 +1,85 @@
+      inherit hash;
+    };
+
+    nativeBuildInputs = [ makeWrapper p7zip ];
```
I haven't tried that myself but wouldn't `makeBinaryWrapper` work as well?

dasJ replied:
Turns out: You don't even need `p7zip` because you already unpacked the zip in the fetcher

leona-ya replied:
Oh yes! `p7z`ip is actually a relict of a really old derivation i tried.

dasJ commented on pkgs/servers/jetbrains/youtrack/default.nix:
```diff
@@ -0,0 +1,85 @@
+
+      makeWrapper "$out/app/bin/youtrack.sh" "$out/bin/youtrack" \
+        --prefix PATH : "$out/libexec/app:${lib.makeBinPath [ jdk17 gawk ]}" \
+        --set FJ_JAVA_EXEC "${jdk17}/bin/java" \
```
Is there anything this fixes? Because I don't set this var and it works perfectly well

dasJ commented on pkgs/servers/jetbrains/youtrack/default.nix:
```diff
@@ -0,0 +1,85 @@
+      done
+
+      makeWrapper "$out/app/bin/youtrack.sh" "$out/bin/youtrack" \
+        --prefix PATH : "$out/libexec/app:${lib.makeBinPath [ jdk17 gawk ]}" \
```
I only add `gawk` to $PATH, no need to add jdk17 and $out/libexec (which doesn't exist in the first place)

dasJ commented on pkgs/servers/jetbrains/youtrack/default.nix:
```diff
@@ -0,0 +1,85 @@
+        --prefix PATH : "$out/libexec/app:${lib.makeBinPath [ jdk17 gawk ]}" \
+        --set FJ_JAVA_EXEC "${jdk17}/bin/java" \
+        --set JRE_HOME ${jdk17}
+
```
I also do `rm -rf $out/internal/java` to get rid of the bundled java

dasJ reviewed (dismissed):
Only looked at the package so far

leona-ya commented:
Finally I now bumped this to 2023.2 and tested it:
* [x] migration from current nixpkgs master 2022.3 -> this branches 2022.3 works
* [x] migration from 2022.3 -> 2023.2 works (with manual interaction)
* [x] migration within 2023.x works without manual interaction

leona-ya commented:
I didn't test the explicit youtrack settings, as something seems cursed with that. The easy `services.youtrack.address`, `services.youtrack.port` settings work.

leona-ya committed:
nixos/youtrack: rebuild module for 2023.x support

leona-ya commented:
Updated now to 2023.3

rvem commented on nixos/modules/services/web-apps/youtrack.nix:
```diff
@@ -1,131 +1,178 @@
+          mkdir -p ${cfg.statePath}/{backups,conf,data,logs,temp}
+        '' + lib.optionalString (lib.versionAtLeast package.version "2023.1") ''
+          ${package}/bin/youtrack configure -J-Ddisable.configuration.wizard.on.upgrade=true
+          ${package}/bin/youtrack configure --listen-port=${toString cfg.port} ${lib.concatStringsSep " " (lib.mapAttrsToList (name: value: "--${name}=${toString value}") cfg.settings )}
```
The port is already a part of `settings`
```suggestion
          ${package}/bin/youtrack configure ${lib.concatStringsSep " " (lib.mapAttrsToList (name: value: "--${name}=${toString value}") cfg.settings )}
```

rvem replied:
Also, there is an issue with `youtrack configure` caused by the fact that the config is actually stateful, so for example, you may add `settings.tls-redirect-from-http = true;` to your configuration, switch to the configuration, remove `settings.tls-redirect-from-http = true;` from your configuration, and switch once again. However, `tls-redirect-from-http=1` will still be present in the config files within `/var/lib/youtrack/conf`

rvem replied:
Sadly, AFAICS, there is no way to specify these settings as a part of `youtrack run`, so perhaps, it's worth cleaning existing configuration altogether before applying `cfg.settings`

leona-ya replied:
That is also my level of information. I'm not sure tbh how to clean the configuration? At least AFAIK there is no file that's only written to by `youtrack confiure` and not by the web ui. 
But also the configuration is the one thing that still confuses me, so happy to accept solutions :)

leona-ya replied:
But also I'll try that out again

rvem replied:
> But also the configuration is the one thing that still confuses me, so happy to accept solutions

Agree that the configuration mechanism is now very weird, and I also didn't yet come up with a way to solve this problem:(

leona-ya replied:
Just looked at what `youtrack configure` does exactly. When using environment parameters, it sets the value all over the `/conf` directory:
```
[root@nixpkgs:/var/lib/youtrack/conf]# rg listen-port
youtrack/service-config.properties
88:bundle.listen-port=8099

internal/services/adminService/service-config.properties
70:bundle.listen-port=8099

hub/service-config.properties
61:bundle.listen-port=8099

internal/bundle.properties
12:listen-port=8099

internal/services/startingPage/service-config.properties
66:bundle.listen-port=8099

internal/services/bundleProcess/service-config.properties
67:bundle.listen-port=8099

internal/services/configurationWizard/service-config.properties
55:jetbrains.bundle.listen-port.default=8099

internal/services/bundleBackend/service-config.properties
66:bundle.listen-port=8099
``` 

but for jvm options, it just sets them in `/conf/youtrack.jvmoptions`. And this file can probably be stored in the nix store.
```
[root@nixpkgs:/var/lib/youtrack/conf]# rg default.page
youtrack.jvmoptions
65:-Djetbrains.youtrack.default.page=/agiles/157-0/current
```

leona-ya replied:
I now did one approach about that. Not 100% good, but much better than before imo. Feel free to review :)

rvem replied:
Looks better now:)

rvem commented on nixos/modules/services/web-apps/youtrack.nix:
```diff
@@ -1,131 +1,178 @@
+      '';
+      example = {
+        tls-redirect-from-http = true;
+        jetbrains.youtrack.admin.restore = true;
```
Looks like options from [general parameter](https://www.jetbrains.com/help/youtrack/server/2023.1/YouTrack-Java-Start-Parameters.html#general-parameters) need to be specified as JVM options, i.e. something like `youtrack configure -J-Djetbrains.youtrack.admin.restore=true`, so this particular example doesn't look correct.
Specifying [environment parameters](https://www.jetbrains.com/help/youtrack/server/2023.1/YouTrack-Java-Start-Parameters.html#environmental-parameters) works just fine :relieved:

rvem reviewed (commented):
I was quite surprised to find out that someone already came up with changes required for the ZIP YouTrack distribution :sweat_smile:

rvem commented on nixos/modules/services/web-apps/youtrack.md:
````diff
@@ -0,0 +1,27 @@
+* JetBrains YouTrack 2023.3 Configuration Wizard will be available on [http://127.0.0.1:8090/?wizard_token=somelongtoken] after start
+```
+
+## Upgrade from 2022.3 to 2023.x {#module-services-youtrack-upgrade-2022_3-2023_1}
````
I had issues with YouTrack DB migration while updating from `2022.3` to `2023.3`.
It was suggested to provide the `-J-Dexodus.entityStore.refactoring.forceAll=true` JVM option for the first YT startup after the upgrade. Perhaps, it's worth mentioning to save someone else time in the future.
However, I couldn't reproduce this on a small YT instance with a free license, so this might only affect some old-ish big YT installations.

leona-ya commented:
Also integrated some more warnings and option migrations now.

leona-ya commented:
Added one more warning about the deprecation of `services.youtrack.extraParams`.

kloenk reviewed (approved):
LGTM, deployed on my aarch64-linux

yu-re-ka commented:
do you want to wait until @dasJ re-reviewed, or can I merge this?

leona-ya commented:
If @dasJ want to review it, I would wait for it :) Otherwise you can probably merge

leona-ya committed:
youtrack_2022_3: init at 2022.3.65371

As the major update of youtrack 2022.3 -> 2023.3 requires manual interaction,
this adds a package pinned to the old version.

leona-ya committed:
youtrack: 2022.3.65371 -> 2023.3.23390

This updates YouTrack to a new major version with a new packaging format.

leona-ya committed:
youtrack: add update script

leona-ya commented:
I moved the packages to `pkgs/by-name` and removed an unnecessary version check in the module
